admin/autoupdater/src/autoupdater.c

```diff
@@ -0,0 +1,458 @@
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+#include "manifest.h"
+#include "settings.h"
+#include "uclient.h"
+#include "util.h"
+#include "version.h"
+
+#include <libplatforminfo.h>
+#include <libubox/uloop.h>
+#include <ecdsautil/ecdsa.h>
+#include <ecdsautil/sha256.h>
+
+#include <fcntl.h>
+#include <getopt.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/types.h>
+#include <sys/file.h>
+#include <sys/stat.h>
+
+
+#define MAX_LINE_LENGTH 512
+#define STRINGIFY(str) #str
+
+static const char *const download_d_dir = "/usr/lib/autoupdater/download.d";
+static const char *const abort_d_dir = "/usr/lib/autoupdater/abort.d";
+static const char *const upgrade_d_dir = "/usr/lib/autoupdater/upgrade.d";
+static const char *const lockfile = "/var/run/autoupdater.lock";
+static const char *const firmware_path = "/tmp/firmware.bin";
+static const char *const sysupgrade_path = "/sbin/sysupgrade";
+
+
+struct recv_manifest_ctx {
+	struct settings *s;
+	struct manifest m;
+	char buf[MAX_LINE_LENGTH + 1];
+	char *ptr;
+};
+
+struct recv_image_ctx {
+	int fd;
+	ecdsa_sha256_context_t hash_ctx;
+};
+
+
+static void usage(void) {
+	fputs("\n"
+		"Usage: autoupdater [options] [<mirror> ...]\n\n"
+		"Possible options are:\n"
+		"  -b, --branch BRANCH  Override the branch given in the configuration.\n\n"
+		"  -f, --force          Always upgrade to a new version, ignoring its priority\n"
+		"                       and whether the autoupdater even is enabled.\n\n"
+		"  -h, --help           Show this help.\n\n"
+		"  -n, --no-upgrade,    Download and validate the manifest as usual, but do not\n"
+		"      --simulate       really flash a new firmware if one is available.\n\n"
+		"  --fallback           Upgrade if and only if the upgrade timespan of the new\n"
+		"                       version has passed for at least 24 hours.\n\n"
+		"  <mirror> ...         Override the mirror URLs given in the configuration. If\n"
+		"                       specified, these are not shuffled.\n\n",
+		stderr
+	);
+}
+
+
+static void parse_args(int argc, char *argv[], struct settings *settings) {
+	enum option_values {
+		OPTION_BRANCH = 'b',
+		OPTION_FORCE = 'f',
+		OPTION_HELP = 'h',
+		OPTION_SIMULATE = 'n',
+		OPTION_FALLBACK = 256,
+	};
+
+	const struct option options[] = {
+		{"branch",     required_argument, NULL, OPTION_BRANCH},
+		{"force",      no_argument,       NULL, OPTION_FORCE},
+		{"fallback",   no_argument,       NULL, OPTION_FALLBACK},
+		{"no-upgrade", no_argument,       NULL, OPTION_SIMULATE},
+		{"simulate",   no_argument,       NULL, OPTION_SIMULATE},
```

Comment:
Maybe. My reasoning: I *hate* when the short option doesn't match the long one. And because `-n` is the canonical short option for "don't really do anything", I had to find something with n. But because `--no-upgrade` didn't sound right (because what does the autoupdater do if not update/upgrade?), I also introduced `--simulate`.

So ideally, someone would propose a `--n*` option that makes more sense than `--no-upgrade` and we drop `--simulate`.